modules/internal/DefaultRectangular.chpl

```diff
@@ -848,32 +1294,127 @@ module DefaultRectangular {
         blk(dim) = blk(dim+1) * dom.dsiDim(dim+1).length;
       computeFactoredOffs();
       var size = blk(1) * dom.dsiDim(1).length;
-      data = _ddata_allocate(eltType, size);
+
+      if defRectSimpleDData {
+        data1 = _ddata_allocate(eltType, size);
+      } else {
+        //
+        // Checking the size first (and having a large-ish size hurdle)
+        // prevents us from calling the pure virtual getChildCount() in
+        // ChapelLocale, when we're setting up arrays in the locale model
+        // and thus here.getChildCount() isn't available yet.
+
+        //
+        if (size < defRectArrMultiDDataSizeThreshold
+            || here.getChildCount() < 2) {
+          mdParDim = 1;
+          mdNumChunks = 1;
+        }
+        else {
+          const (numChunks, parDim) =
+            _computeChunkStuff(here.getChildCount(),
+                               true /* ignoreRunning */,
+                               1 /* minIndicesPerTask */,
+                               dom.ranges);
+          if numChunks == 0 {
+            mdParDim = 1;
+            mdNumChunks = 1;
+          } else {
+            mdParDim = parDim;
+            mdNumChunks = numChunks;
+          }
+        }
+        mdRLo = dom.dsiDim(mdParDim).alignedLow;
+        mdRHi = dom.dsiDim(mdParDim).alignedHigh;
+        mdRLen = dom.dsiDim(mdParDim).length;
+        dataVec = _ddata_allocate(_ddata(eltType), mdNumChunks);
+        for i in 0..#mdNumChunks {
```

Comment:
Can we add a comment or TODO here describing the future intention of making sure we allocate in a locality-aware fashion?